components/webext-storage/src/store.rs

```diff
@@ -110,18 +116,32 @@ impl Store {
     /// Returns the bytes in use for the specified items (which can be null,
     /// a string, or an array)
     pub fn get_bytes_in_use(&self, ext_id: &str, keys: JsonValue) -> Result<usize> {
-        api::get_bytes_in_use(&self.db, ext_id, keys)
+        let db = self.db.lock();
+        api::get_bytes_in_use(&db, ext_id, keys)
     }
 
     /// Returns a bridged sync engine for Desktop for this store.
-    pub fn bridged_engine(&self) -> sync::BridgedEngine<'_> {
+    pub fn bridged_engine(&self) -> sync::BridgedEngine {
         sync::BridgedEngine::new(&self.db)
     }
 
     /// Closes the store and its database connection. See the docs for
     /// `StorageDb::close` for more details on when this can fail.
-    pub fn close(self) -> result::Result<(), (Store, Error)> {
-        self.db.close().map_err(|(db, err)| (Store { db }, err))
+    pub fn close(self) -> Result<()> {
+        // Even though this consumes `self`, the fact we use an Arc<> means
+        // we can't guarantee we can actually consume the inner DB - so do
+        // the best we can.
+        let shared: SharedStorageDb = match Arc::try_unwrap(self.db) {
+            Ok(shared) => shared,
+            _ => {
+                // Probably means the sync engine remains alive?
+                log::warn!("Attempting to close a store while other DB references exist.");
```

Comment:
Hmmm what do you think about implementing `Drop` on the `SharedStorageDb`? That way whenever the underlying memory is dropped, we close the actual connection, regardless of when that happens... which should be **soon** after this `close` function runs... thoughts?


But maybe this `close` function could also be a `Drop` impl on the `Store`, but I'm not sure if we want to close a connection without dropping the whole store so that's another problem

Reply:
> whenever the underlying memory is dropped, we close the actual connection, regardless of when that happens...

I think the tricky part is that `close()` can fail. Rusqlite has something similar, and panics if it can't close a dropped connection, which unfortunately turns into a shutdown crash in Desktop because of `panic = "abort"`. (For comparison, `std::fs::File` is a little kinder, and ignores errors in its `Drop` implementation).

Surfacing `close()` errors on Desktop lets `AsyncShutdown.jsm` make some logging noise while still unblocking shutdown, and leaking the `StorageDb` avoids the shutdown crash.

Reply:
eeek this is the comment I missed 😵‍💫 yeah this is getting tricky - thanks @linabutler 🙏 🙇‍♂️

Reply:
So I think the resolution to this and the other massive thread is that:
* We hope/pray that the consumers somehow manage to avoid contention here, and we believe desktop does via its uses of the task queue.
* But just in-case, this code remains the same (ie, we never panic if we aren't the exclusive arc holder.)
* And just in-double-case, we implement `Drop` on the inner `StorageDb` as a fail-safe so we never implicitly panic as the connection drops.

I'll update the patch accordingly.

Reply:
Sounds perfect, thanks!